### DISCUSSION TITLE
$where is not allowed in this context, but request correctly.

### DISCUSSION THREAD

**Discussion by Qu1eeeOJ** [jenssegers/laravel-mongodb#2625] [Q&A] [answered]
Hello, I wrote such a request using laravel ORM.

If you insert this request directly into mongodb, then the request is correct, but in laravel it gives an error: $where is not allowed in this context

There my ORM Request:
```
Item::query()->select('account_id', 'item_id', 'offer_id', 'fbo_sku', 'fbs_sku', 'barcodes')
  ->orWhere('item_id', ''like,  "%{$searchable}%")
  ->orWhere('name', 'like', "%{$searchable}%")
  ->orWhereRaw(['$where' => "/^{$searchable}.*/.test(this.item_id)"])
  ->orWhereRaw(['$where' => "/^{$searchable}.*/.test(this.fbo_sku)"])
  ->orWhereRaw(['$where' => "/^{$searchable}.*/.test(this.fbs_sku)"])
  ->get();
```

Exception:
```
MongoDB\Driver\Exception\CommandException {#1421 // app/Services/DataTables/DataTableService.php:138
  #message: "$where is not allowed in this context"
  #code: 2
  #file: "/var/www/project/vendor/mongodb/mongodb/src/Operation/Aggregate.php"
  #line: 379
  #errorLabels: null
  #resultDocument: {#1422
    +"ok": 0.0
    +"errmsg": "$where is not allowed in this context"
    +"code": 2
    +"codeName": "BadValue"
  }

```

**Comment by GromNaN** [comment]
I just tried the provided request and didn't get any error. I used on version 4.0.0-rc1, what is your version?

Alternatively you should use `MongoDB\BSON\Regex` objects to perform regex queries. You could build the query like this instead of using [`$where`](https://www.mongodb.com/docs/manual/reference/operator/query/where/).
```php
use MongoDB\BSON\Regex;
Item::query()->select('account_id', 'item_id', 'offer_id', 'fbo_sku', 'fbs_sku', 'barcodes')
  ->orWhere('item_id', new Regex(preg_quote($searchable), 'i'))
  ->orWhere('name', new Regex(preg_quote($searchable), 'i'))
  ->orWhere('fbo_sku', new Regex('^'.preg_quote($searchable), ''))
  ->orWhere('fbs_sku', new Regex('^'.preg_quote($searchable), ''))
  ->get();
```

**Reply by Qu1eeeOJ** [reply]
Yes, I'm using 4.0.0-rc1.

The full code:
```php
$searchable = "514669594";

        $query = Item::query()->select(['account_id', 'item_id', 'offer_id', 'fbo_sku', 'fbs_sku', 'barcodes', 'primary_image', 'old_price', 'marketing_price', 'price', 'status', 'name'])
            ->orWhere('name', 'like', "%{$searchable}%")
            ->orWhere('offer_id', 'like', "%{$searchable}%")
            ->orWhere('barcodes', 'like', "%{$searchable}%")
            ->orWhereRaw(['$where' => "/^{$searchable}.*/.test(this.item_id)"])
            ->orWhereRaw(['$where' => "/^{$searchable}.*/.test(this.fbo_sku)"])
            ->orWhereRaw(['$where' => "/^{$searchable}.*/.test(this.fbs_sku)"]);

        return response()->json([
            'data' => $query->get(),
            'count' => $query->count()
        ]);
```

**Reply by GromNaN** [reply]
It's the `count` query that fails.
Here is the generated call to the MongoDB library:

```php
$collection->countDocuments([
    '$or' => [
        [ 'name' => new Regex('^.*514669594.*$', 'i') ],
        [ 'offer_id' => new Regex('^.*514669594.*$', 'i') ],
        [ 'barcodes' => new Regex('^.*514669594.*$', 'i') ],
        [ '$where' => '/^514669594.*/.test(this.item_id)' ],
        [ '$where' => '/^514669594.*/.test(this.fbo_sku)' ],
        [ '$where' => '/^514669594.*/.test(this.fbs_sku)' ],
    ],
];
```

This query is converted into an aggregation [by the library](https://github.com/mongodb/mongo-php-library/blob/ec6c431acc87ced5e85a163b92e76a8d8f69e000/src/Operation/CountDocuments.php#L145-L157).

```
$collection->aggregates([
    '$match' => [
        '$or' => [
            [ 'name' => new Regex('^.*514669594.*$', 'i') ],
            [ 'offer_id' => new Regex('^.*514669594.*$', 'i') ],
            [ 'barcodes' => new Regex('^.*514669594.*$', 'i') ],
            [ '$where' => '/^514669594.*/.test(this.item_id)' ],
            [ '$where' => '/^514669594.*/.test(this.fbo_sku)' ],
            [ '$where' => '/^514669594.*/.test(this.fbs_sku)' ],
        ],
     ],
     '$group' => [
         '_id' => 1,
         'n' => ['$sum' => 1],
     ],
];
```

I just tried the query with mongosh, it also fails with the same error message.

```
MongoServerError: $where is not allowed in this context
```

You can follow this Jira ticket if you want to be informed when we will work on this issue: [DRIVERS-2728](https://jira.mongodb.org/browse/DRIVERS-2728)

**Reply by GromNaN** [reply]
My recommendation would be to rewrite your query using a Regex as mentioned in the [previous comment](https://github.com/mongodb/laravel-mongodb/discussions/2625?notification_referrer_id=NT_kwDOAAYaorE3NzY3MzkyNDk2OjQwMDAzNA&notifications_query=org%3Amongodb#discussioncomment-7046290).

I don't think `$where` will ever be supported in aggregation pipelines.

**Reply by Qu1eeeOJ** [reply]
I'm sorry, but how can I now correctly count documents by integer fields?

Will your answer above work for non-number type fields?

**Reply by GromNaN** [reply]
I don't understand your question. Please provide an example of your data and the expected results, and how you are trying to do.

**Reply by Qu1eeeOJ** [reply]
I have the fields fbo_sky, fbs_sky - these fields are of type int32, then how can I now count documents using regular expressions on them?

**Reply by GromNaN** [reply, marked as answer]
You need to create a raw aggregation and group by this fields. The type of this fields doesn't matter.

https://www.mongodb.com/docs/manual/reference/operator/aggregation/count/#behavior

With Laravel-mongodb, use the Model::raw()->aggregate() method.

**Reply by Qu1eeeOJ** [reply]
Understood, thank you very much for your help!
